### PR TITLE
ENH: Implement NDC from Yeh2019

### PR DESCRIPTION
**Pull request by mattcieslak**
Adds a function to address #3154.

What would you recommend to add for a test?

### CODE REVIEW

File: dipy/denoise/qc.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+        vectorized_images = [
+            dwi_data[..., idx][mask > 0] for idx in dwi_indices]
+    else:
+        vectorized_images = [
+            dwi_data[..., idx].flatten() for idx in dwi_indices]
+
+    # Get a pseudo-qspace value for b>0s
+    qvecs = (np.sqrt(gtab.bvals)[:, np.newaxis] * gtab.bvecs)[dwi_mask]
+
+    neighbor_correlations = []
```
Comment (arokem) [inline dipy/denoise/qc.py:50]: Couldn't this be preallocated for speed?

**Review by arokem — COMMENTED** [review]
Looks good! Regarding testing, I'd suggest testing that correlation between a signal and itself is 1, maybe with different sets of b-vectors and b-values. Maybe also that this correlation doesn't change if one of the signals is scaled. For smoke testing, worth testing with different gradient schemes, with and without b=0, for example.

File: dipy/denoise/qc.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
+import numpy as np
+from dipy.core.geometry import cart_distance
+
+
+def neighboring_dwi_correlation(dwi_data, gtab, mask=None):
+    r"""Calculate the Neighboring DWI Correlation (NDC) from dMRI data.
```
Comment (skoudoro) [inline dipy/denoise/qc.py:6]: `r` not needed. triple double quotes should be enough

File: dipy/denoise/qc.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    correlations. According to [Yeh2019], an NDC less than 0.4 indicates a
+    low quality image.
+
+    Parameters
+    ----------
+    dwi_data : 4D ndarray
+        dwi data on which to calculate NDC
+    gtab : dipy.core.gradients.GradientTable
+        Gradient table.
+    mask : 3D ndarray
```
Comment (skoudoro) [inline dipy/denoise/qc.py:18]: `3D ndarray, optional` instead

**Review by skoudoro — COMMENTED** [review]
Also there is no automatic module discovery.

So, this new file needs to be added in [`meson.build`](https://github.com/dipy/dipy/blob/master/dipy/denoise/meson.build).

Thank you for this. Personally, I need to read the paper

File: dipy/denoise/qc.py (hunk 1 above)
Comment (mattcieslak) [reply dipy/denoise/qc.py]: this is only written to and read once, so I'd favor a list for readability

Comment (arokem) [reply dipy/denoise/qc.py]: Fair enough

**Comment by mattcieslak** [comment]
should be good for review!

**Comment by arokem** [comment]
LGTM!

**Comment by ShreyasFadnavis** [comment]
Looks great! Thanks for this @mattcieslak ! -- Not sure if `denoise` is the right module for it though? 

I don't want people to miss this utility, just because they did not look in the right place.

Alternative modules could be `stats`?

**Comment by skoudoro** [comment]
Is there any additional comment @arokem @ShreyasFadnavis or @araikes?? 

I think this PR is ready to get in.

**Comment by skoudoro** [comment]
Thank you for this @mattcieslak !

**Pull request merged**